xla/service/gpu/gpu_compiler.cc

```diff
@@ -487,10 +487,14 @@ Status GpuCompiler::OptimizeHloModule(HloModule* hlo_module,
     pipeline.AddPass<TopkDecomposer>();
 
     HloPredicate upcaster_filter = [&](const HloInstruction* instr) {
+#if GOOGLE_CUDA            
```

Comment:
Can you please create virtual method for this, and implement it differently in amdgpu_compiler and nvptx_compiler? Similar to how it was done in https://github.com/openxla/xla/commit/eff2b737cb4f6949fa85e3646eed3db684826dfc

Reply:
Much apprecipate for the example. I will look it.

Reply:
Just want to make sure in case I misunderstood here. Do you expect to have two functions somehow like this:

```
    TF_RETURN_IF_ERROR(AddOperandUpcasterPasses(
      &pipeline, hlo_module, stream_exec, debug_options, options,
      gpu_target_config, upcaster_filter));

    TF_RETURN_IF_ERROR(AddResultCasterPasses(
      &pipeline, hlo_module, stream_exec, debug_options, options,
      gpu_target_config, upcaster_filter));
```

Then, these two virtual functions will be implemented differently in amdgpu_compiler and nvptx_compiler, respectively. The only difference I think might just this one:
```
upcaster_filter = [&](const HloInstruction* instr) {
      return !stream_exec->GetDeviceDescription()
                  .cuda_compute_capability()
                  .IsAtLeast(se::CudaComputeCapability::VOLTA) ||
             !gpu::IsMatrixMultiplication(*instr);
}
```

If the purpose of doing the above is to remove the annoying preprocessor #if #else things, I can use gpu_target_config.platform_name or gpu_version to determine it at the runtime.

Like this 
```
      if (gpu_target_config.platform_name == "ROCM"){
        return !gpu::IsMatrixMultiplication(*instr);
      }
```

Reply:
Thanks, determining it at runtime is much better :)

Reply:
Done.